src/resources/plugin/index.ts

```diff
@@ -0,0 +1,100 @@
+import type {
+  Plugin,
+  ResourceStoreContext,
+  RouterContext,
+  RouteResourceResponse,
+} from '../../common/types';
+import { getResourceStore } from '../../controllers/resource-store';
+import { getResourcesForNextLocation } from '../../controllers/resource-store/utils';
+import { getRouterState } from '../../controllers/router-store';
+
+const loadOnUrlChange = (
+  context: RouterContext,
+  prevContext: RouterContext
+) => {
+  const { requestResources, getContext: getResourceStoreContext } =
+    getResourceStore().actions;
+
+  const nextResources = getResourcesForNextLocation(
+    prevContext,
+    context,
+    getResourceStoreContext()
+  );
+
+  return Promise.all(requestResources(nextResources, context, {}));
+};
+
+const beforeLoad = ({
+  context,
+  nextContext,
+}: {
+  context: RouterContext;
+  nextContext: RouterContext;
+}) => {
+  const { cleanExpiredResources, getContext: getResourceStoreContext } =
+    getResourceStore().actions;
+  const nextResources = getResourcesForNextLocation(
+    context,
+    nextContext,
+    getResourceStoreContext()
+  );
+  cleanExpiredResources(nextResources, nextContext);
+};
+
+export const createResourcesPlugin = ({
+  context: initialResourceContext,
+  resourceData: initialResourceData,
+  timeout,
+}: {
+  context?: ResourceStoreContext;
+  resourceData: any;
+  timeout?: number;
+}): Plugin<{
+  resources: Promise<RouteResourceResponse<unknown>[]>;
+}> => {
+  return {
+    onHydrate: () => {
+      getResourceStore().actions.hydrate({
+        resourceContext: initialResourceContext,
+        resourceData: initialResourceData,
+      });
+    },
+    onBeforeRouteLoad: beforeLoad,
+    onRouteLoad: ({ context, prevContext }) => {
+      const { route, match, query } = context;
+      // TODO: in next refactoring add `if (route.resources)` check
+      // For now requesting resources for every route even if `resources` prop is missing on Route
+      if (prevContext) {
+        return {
+          resources: loadOnUrlChange(context, prevContext),
+        };
+      }
+
+      return {
+        resources: getResourceStore().actions.requestAllResources(
+          {
+            route,
+            match,
+            query,
+          },
+          { timeout }
+        ),
+      };
+    },
+    onRoutePrefetch: (context: RouterContext) => {
+      const { route, match, query } = getRouterState();
```

Comment:
Isn't this data available on `context: RouterContext` ? `getRouterState` is a sort of private API, so if possible I'd like plugins to not reach into it